src/graph.rs

```diff
@@ -86,6 +92,131 @@ impl<'tree> Graph<'tree> {
+            {
+                let sink = self.0;
+                let edge = self.1;
+                let ctx = self.2;
+                let mut map = serializer.serialize_map(None)?;
+                map.serialize_entry("sink", &sink)?;
+                if !edge.attributes.values.is_empty() {
+                    map.serialize_entry("attrs", &JSONAttributes(&edge.attributes, &ctx))?;
+                }
+                map.end()
+            }
+        }
+
+        impl<'a> ser::Serialize for JSONAttributes<'a> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let attrs = self.0;
+                let ctx = self.1;
+                let mut map = serializer.serialize_map(None)?;
+                for (_, (key, value)) in attrs.values.iter().enumerate() {
+                    map.serialize_entry(&JSONIdentifier(key, ctx), &JSONValue(value, ctx))?;
+                }
+                map.end()
+            }
+        }
+
+        impl<'a> ser::Serialize for JSONIdentifier<'a> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let identifier = self.0;
+                let ctx = self.1;
+                let symbol = ctx.identifiers.resolve(identifier.0).unwrap();
+                serializer.serialize_str(symbol)
+            }
+        }
+
+        impl<'a> ser::Serialize for JSONValue<'a> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let value = self.0;
+                let ctx = self.1;
+                match value {
+                    Value::Null => serializer.serialize_none(),
+                    Value::Boolean(bool) => serializer.serialize_bool(*bool),
+                    Value::Integer(integer) => serializer.serialize_u32(*integer),
+                    Value::String(string) => serializer.serialize_str(string),
+                    // FIXME: there's no way to distinguish sets and lists, so we can't roundtrip accurately
+                    Value::List(list) => {
+                        serializer.collect_seq(list.iter().map(|value| JSONValue(value, ctx)))
+                    }
+                    Value::Set(set) => {
+                        serializer.collect_seq(set.iter().map(|value| JSONValue(value, ctx)))
+                    }
+                    // FIXME: we don't distinguish between syntax tree node IDs, graph node IDs, and integers
+                    Value::SyntaxNode(node) => serializer.serialize_u32(node.0),
+                    Value::GraphNode(node) => serializer.serialize_u32(node.0),
+                }
+            }
+        }
+
+        let json_graph = JSONGraph(self, ctx);
+        let s = serde_json::to_string_pretty(&json_graph).unwrap();
```

Comment:
I think this should probably be a `Serialize` implementation for `Graph`, but in the meantime we can get readable JSON out of it.